wrench/src/wrench.rs

```diff
@@ -0,0 +1,361 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use app_units::Au;
+#[cfg(windows)]
+use dwrote;
+use euclid::Size2D;
+use gleam::gl;
+use glutin;
+use glutin::WindowProxy;
+use image;
+use image::GenericImage;
+use std::collections::HashMap;
+use std::ffi::CStr;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use webrender;
+use webrender_traits::*;
+use yaml_rust::Yaml;
+use yaml_frame_writer::YamlFrameWriter;
+
+use {WHITE_COLOR, BLACK_COLOR};
+
+struct Notifier {
+    window_proxy: WindowProxy,
+}
+
+impl Notifier {
+    fn new(window_proxy: WindowProxy) -> Notifier {
+        Notifier {
+            window_proxy: window_proxy,
+        }
+    }
+}
+
+impl RenderNotifier for Notifier {
+    fn new_frame_ready(&mut self) {
+        self.window_proxy.wakeup_event_loop();
+    }
+
+    fn new_scroll_frame_ready(&mut self, _composite_needed: bool) {
+        self.window_proxy.wakeup_event_loop();
+    }
+
+    fn pipeline_size_changed(&mut self,
+                             _: PipelineId,
+                             _: Option<Size2D<f32>>) {
+    }
+}
+
+#[cfg(target_os = "windows")]
+pub fn layout_simple_ascii(face: NativeFontHandle, text: &str, size: Au) -> (Vec<u16>, Vec<f32>) {
+    let system_fc = dwrote::FontCollection::system();
+    let font = system_fc.get_font_from_descriptor(&(face as dwrote::FontDescriptor)).unwrap();
+    let face = font.create_font_face();
+
+    let chars: Vec<u32> = text.chars().map(|c| c as u32).collect();
+    let indices = face.get_glyph_indices(&chars);
+    let glyph_metrics = face.get_design_glyph_metrics(&indices, false);
+
+    let device_pixel_ratio: f32 = 1.0;
+    let em_size = size.to_f32_px() / 16.;
+    let design_units_per_pixel = face.metrics().designUnitsPerEm as f32 / 16. as f32;
+    let scaled_design_units_to_pixels = (em_size * device_pixel_ratio) / design_units_per_pixel;
+
+    let advances = glyph_metrics.iter().map(|m| m.advanceWidth as f32 * scaled_design_units_to_pixels).collect();
+
+    (indices, advances)
+}
+
+#[cfg(not(target_os = "windows"))]
+pub fn layout_simple_ascii(face: NativeFontHandle, text: &str, size: Au) -> (Vec<u16>, Vec<f32>) {
+    panic!("Can't layout simple ascii on this platform");
+}
+
+pub trait WrenchThing {
+    fn next_frame(&mut self);
+    fn prev_frame(&mut self);
+    fn do_frame(&mut self, &mut Wrench) -> u32;
+}
+
+pub struct Wrench {
+    pub window: glutin::Window,
```

Comment:
does it all have to be public?
I know this is not much important, given that it's not a used API but just a tool
wondering just out of interest